core/state_transition.go

```diff
@@ -163,12 +165,64 @@ func IntrinsicGas(data []byte, accessList types.AccessList, isContractCreation b
 		gas += z * params.TxDataZeroGas
 	}
 	if accessList != nil {
+		accessListGas, err := AccessListGas(rules, accessList)
+		if err != nil {
+			return 0, err
+		}
+		totalGas, overflow := commonMath.SafeAdd(gas, accessListGas)
+		if overflow {
+			return 0, fmt.Errorf("overflow adding access list gas")
+		}
+		gas = totalGas
+	}
+
+	return gas, nil
+}
+
+func AccessListGas(rules params.Rules, accessList types.AccessList) (uint64, error) {
+	var gas uint64
+	if !rules.PredicatesExist() {
 		gas += uint64(len(accessList)) * params.TxAccessListAddressGas
 		gas += uint64(accessList.StorageKeys()) * params.TxAccessListStorageKeyGas
+		return gas, nil
 	}
+
+	for i, accessTuple := range accessList {
+		address := accessTuple.Address
+		if !rules.PredicateExists(address) {
+			totalGas, overflow := commonMath.SafeAdd(gas, params.TxAccessListAddressGas+uint64(len(accessTuple.StorageKeys))*params.TxAccessListStorageKeyGas)
+			if overflow {
```

Comment:
nit: can this be on its own line? its a bit hard to read `params.TxAccessListAddressGas+uint64(len(accessTuple.StorageKeys))*params.TxAccessListStorageKeyGas)`

Reply:
also can this addition ever overflow ^

Reply:
Good call, moving to a separate line and adding a comment